### PR TITLE
Implement __array__ on ExtensionIndex

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/tests/indexes/period/test_period.py [hunk 1]
```diff
@@ -677,3 +677,32 @@ def test_is_monotonic_with_nat():
         assert not obj.is_monotonic_increasing
         assert not obj.is_monotonic_decreasing
         assert obj.is_unique
+
+
+@pytest.mark.parametrize("array", [True, False])
+def test_dunder_array(array):
```
Comment (jreback) [inline pandas/tests/indexes/period/test_period.py]: do we have this test for indexes generally?

**Comment by jreback** [comment]
looks ok. @TomAugspurger if you'd have a look.

**Comment by jreback** [comment]
do we have an issue for adding ``__array_ufunc__`` to index?

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -90,6 +90,8 @@ Backwards incompatible API changes
   now raise a  ``TypeError`` if a not-accepted keyword argument is passed into it.
   Previously a ``UnsupportedFunctionCall`` was raised (``AssertionError`` if ``min_count`` passed into :meth:`~DataFrameGroupby.median``) (:issue:`31485`)
 - :meth:`DataFrame.at` and :meth:`Series.at` will raise a ``TypeError`` instead of a ``ValueError`` if an incompatible key is passed, and ``KeyError`` if a missing key is passed, matching the behavior of ``.loc[]`` (:issue:`31722`)
+- Passing an integer dtype other than ``int64`` to ``np.array(period_index, dtype=...)`` will now raise ``NotImplementedError`` instead of incorrectly using ``int64`` (:issue:`32255`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v1.1.0.rst]: Why NotImplementedError, rathan something like ValueError? Are we the ones raising it, or is NumPy?

Comment (TomAugspurger) [reply doc/source/whatsnew/v1.1.0.rst]: Though from the tests, it looks like a TypeError is raised (unless I'm misreading).

**Review by TomAugspurger — COMMENTED** [review]
Don't see an issue for `Index.__array_ufunc__`. Opened https://github.com/pandas-dev/pandas/issues/32411 as a placeholder.

**Review by jreback — APPROVED** [review]
lgtm. merge on @TomAugspurger comments

File: pandas/core/arrays/period.py [hunk 3]
```diff
@@ -282,7 +282,12 @@ def freq(self):
         return self.dtype.freq
 
     def __array__(self, dtype=None) -> np.ndarray:
-        # overriding DatetimelikeArray
+        if dtype == "i8":
+            return self.asi8
+        elif dtype == bool:
+            return ~self._isnan
+
+        # This will raise TypeErorr for non-object dtypes
```
Comment (simonjayhawkins) [inline pandas/core/arrays/period.py]: typo for follow-up

**Comment by simonjayhawkins** [comment]
Thanks @jbrockmendel

**Pull request merged**